src/server/scripts/OutdoorPvP/OutdoorPvPNA.cpp

```diff
@@ -82,23 +96,36 @@ TeamId OPvPCapturePointNA::GetControllingFaction() const
     return m_ControllingFaction;
 }
 
-void OPvPCapturePointNA::SpawnNPCsForTeam(TeamId teamId)
+void OPvPCapturePointNA::DeSpawnNPCs(HalaaNPCS teamNPC)
 {
-    const creature_type* creatures = nullptr;
-    if (teamId == TEAM_ALLIANCE)
-        creatures = AllianceControlNPCs;
-    else if (teamId == TEAM_HORDE)
-        creatures = HordeControlNPCs;
-    else
-        return;
-    for (int i = 0; i < NA_CONTROL_NPC_NUM; ++i)
-        AddCreature(i, creatures[i].entry, creatures[i].map, creatures[i].x, creatures[i].y, creatures[i].z, creatures[i].o, 1000000);
+    for (int i = 0; i < 20; i++)
+    {
+        ObjectGuid::LowType spawnId = teamNPC[i];
+        auto bounds = m_PvP->GetMap()->GetCreatureBySpawnIdStore().equal_range(spawnId);
+        const CreatureData* data = sObjectMgr->GetCreatureData(spawnId);
+        for (auto itr = bounds.first; itr != bounds.second;)
+        {
+            // can happen when closing the core
+            Creature* c = itr->second;
+            ++itr;
+            c->AddObjectToRemoveList();
+            sObjectMgr->RemoveCreatureFromGrid(spawnId, data);
+            m_Creatures[i] = 0;
+            m_CreatureTypes[m_Creatures[i]] = 0;
+        }
+    }
 }
 
-void OPvPCapturePointNA::DeSpawnNPCs()
+void OPvPCapturePointNA::SpawnNPCsForTeam(HalaaNPCS teamNPC)
 {
-    for (int i = 0; i < NA_CONTROL_NPC_NUM; ++i)
-        DelCreature(i);
+    for (int i = 0; i < 20; i++)
+    {
+        ObjectGuid::LowType spawnId = teamNPC[i];
+        const CreatureData* data = sObjectMgr->GetCreatureData(spawnId);
```

Comment:
CreatureData const*, and also missing nullcheck for data, it can return null in certain cases.